annotations/src/main/java/org/corfudb/runtime/object/ICorfuSMR.java

```diff
@@ -8,40 +8,60 @@
  * @param <T> The type of the underlying object.
  * Created by mwei on 11/10/16.
  */
-public interface ICorfuSMR<T> {
+public interface ICorfuSMR<T>
+        extends ICorfuExecutionContext<T>, ICorfuVersionPolicy, AutoCloseable {
 
     /** The suffix for all precompiled SMR wrapper classes. */
     String CORFUSMR_SUFFIX = "$CORFUSMR";
 
     /** Get the proxy for this wrapper, to manage the state of the object.
      * @return The proxy for this wrapper. */
-    ICorfuSMRProxy<T> getCorfuSMRProxy();
+    default ICorfuSMRProxy<T> getCorfuSMRProxy() {
+        throw new IllegalStateException("");
+    }
 
     /** Set the proxy for this wrapper, to manage the state of the object.
      * @param proxy The proxy to set for this wrapper. */
-    void setCorfuSMRProxy(ICorfuSMRProxy<T> proxy);
+    default void setCorfuSMRProxy(ICorfuSMRProxy<T> proxy) {
+        throw new IllegalStateException("");
+    }
 
     /** Get a map from strings (function names) to SMR upcalls.
      * @return The SMR upcall map. */
-    Map<String, ICorfuSMRUpcallTarget<T>> getCorfuSMRUpcallMap();
+    default Map<String, ICorfuSMRUpcallTarget<T>> getCorfuSMRUpcallMap() {
+        throw new IllegalStateException("");
+    }
 
     /** Get a map from strings (function names) to undo methods.
      * @return The undo map. */
-    Map<String, IUndoFunction<T>> getCorfuUndoMap();
+    default Map<String, IUndoFunction<T>> getCorfuUndoMap() {
+        throw new IllegalStateException("");
+    }
 
     /** Get a map from strings (function names) to undoRecord methods.
      * @return The undo record map. */
-    Map<String, IUndoRecordFunction<T>> getCorfuUndoRecordMap();
+    default Map<String, IUndoRecordFunction<T>> getCorfuUndoRecordMap() {
```

Comment:
why you need to make all these methods default and throw an exception at the same time. You can just get rid of `default` modifier and then you don't need to throw exceptions

Reply:
The implementation has to be defined this way since the real implementation is generated by Corfu Compile Proxy. If you leave these function unimplemented, you are forcing all SMR objects to define these functions.